### DISCUSSION TITLE
Adding Typescript to a module

### DISCUSSION THREAD

**Discussion by johndcollins** [OrchardCMS/OrchardCore#13685] [Q&A]
Hi,
I'm trying to add Typescript to a module and I'm running into some issues.
I've looked at some of the other modules to see how they do it, but I'm not exactly sure what I'm doing wrong.

It seems there are two ways to control the compiler options. 
1. add a tsconfig.json file in the root of the project
2. add tsCompilerOptions to the assets.json file

I'm not sure what is best.
My files are getting converted to js, but there is some code added to the js file which seems to be messing up my project:

...
Object.defineProperty(exports, "__esModule", {
  value: true
});
...

This is giving me an error: exports is not defined

I've tried all the suggestions here:
https://bobbyhadz.com/blog/typescript-uncaught-referenceerror-exports-is-not-defined#nodejs---referenceerror-exports-is-not-defined

What is the correct way to add typescript to a module and get it working?

**Comment by MichaelPetrinolis** [comment]
if you want to build a small typescript snippet, have Intellisense in vs and copy it to the wwwroot folder, use tsCompilerOptions. If you have a big typescript project, then create a folder, put your sources, package.json, your build pipeline etc. and use the "copy" assets function to copy the output of the build pipeline you created in the ts project. Check OrchardCore.Liquid -> Assets -> monaco -> liquid-intellisense.ts for an example

**Reply by johndcollins** [reply]
Thanks Michael.

Thanks for the example. I was using a couple of the other projects as examples like the GraphQL.
These examples are all one file TypeScript and I'm using multiple files.
My problem is that I don't have much experience with writing javascript, I'm a C# developer. Typescript seems to work better for me when I have to structure my code in classes and files.

The problem I'm running into is when I have multiple TypeScript files and some files depend on others. TypeScript uses the import function to get the other files. When I convert my files to js using the output of the Asset file, the site errors saying import is not valid. How do I get around this and how do I call the first module from a page?

Sorry if this is basic knowledge, but I've been trying to get this to work for a couple of weeks and I just can't seem to get to a solution.
